---
title: "DPOT combined with U16S : ZymoBIOMICS and BEI HM-782D Microbial Community Standard"
author: "Chieh Hua Lin"
date: "2023/7/7"
output: html_document
---

email: [mammer\@nhri.edu.tw](mailto:mammer@nhri.edu.tw){.email}

### Data

In this study, we utilized two mock communities, namely ZymoBIOMICS D6305 (referred to as Zymo8) and BEI HM-782D (referred to as HM20e). The sequencing datasets corresponding to these mock communities were acquired from the NCBI Sequence Read Archive (SRR9089357, SRR5593316).

These mock communities encompass the following bacteria:

| **Zymo8**                     | **HM20e**                                                        |
|------------------------|------------------------------------------------|
| Bacillus subtilis             | Acinetobacter baumannii                                          |
| Enterococcus faecalis         | Actinomyces odontolyticus (current name : Schaalia odontolytica) |
| Escherichia coli              | Bacillus cereus                                                  |
| Limosilactobacillus fermentum | Bacteroides vulgatus (current name: Phocaeicola vulgatus)        |
| Listeria monocytogenes        | Clostridium beijerinckii                                         |
| Pseudomonas aeruginosa        | Deinococcus radiodurans                                          |
| Salmonella enterica           | Enterococcus faecalis                                            |
| Staphylococcus aureus         | Escherichia coli                                                 |
|                               | Helicobacter pylori                                              |
|                               | Lactobacillus gasseri                                            |
|                               | Listeria monocytogenes                                           |
|                               | Neisseria meningitidis                                           |
|                               | Propionibacterium acnes (current name: Cutibacterium acnes)      |
|                               | Pseudomonas aeruginosa                                           |
|                               | Rhodobacter sphaeroides (current name: Cereibacter sphaeroides)  |
|                               | Staphylococcus aureus                                            |
|                               | Staphylococcus epidermidis                                       |
|                               | Streptococcus agalactiae                                         |
|                               | Streptococcus mutans                                             |
|                               | Streptococcus pneumoniae                                         |

#### Please be aware that the subsequent species share identical sequences with those in the mock communities. Consequently, these species were also considered as correct assignments in this analysis.

| **Mock species**         | **Species with identical sequences** |
|--------------------------|--------------------------------------|
| Bacillus cereus          | Bacillus anthracis                   |
|                          | Bacillus mobilis                     |
|                          | Bacillus sp. FDAARGOS_235            |
|                          | Bacillus sp. HBCD-sjtu               |
|                          | Bacillus thuringiensis               |
| Bacillus subtilis        | Alkalihalobacillus gibsonii          |
|                          | Bacillus amyloliquefaciens           |
|                          | Bacillus licheniformis               |
|                          | Bacillus sp. 275                     |
|                          | Bacillus sp. FJAT-14266              |
|                          | Bacillus sp. Lzh-5                   |
|                          | Bacillus sp. MD-5                    |
|                          | Bacillus sp. Pc3                     |
|                          | Bacillus sp. YP1                     |
|                          | Bacillus spizizenii                  |
|                          | Bacillus vallismortis                |
| Clostridium beijerinckii | Clostridium sp. MF28                 |
| Escherichia coli         | Shigella flexneri                    |
|                          | Shigella sonnei                      |
|                          | Shigella sp. PAMC 28760              |
| Lactobacillus iners      | Amylolactobacillus amylotrophicus    |
| Staphylococcus aureus    | Staphylococcus sp. NBRC 15913        |
| Streptococcus agalactiae | Streptococcus sp. 'group B'          |

### 16S rRNA reference database : U16S

U16S is a consolidated, non-redundant 16S rRNA sequence database, born out of the integration of three extensively utilized databases---RefSeq 16S, SILVA, and EzBioCloud---underpinned by methodical taxonomic curation and quality filtering.

## Demostration of DPOT with U16S

```{r setup, include = FALSE}
require("knitr")
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_knit$set(root.dir = "D:/System_bio/Microbiome/Manuscript/")
```

```{r echo=TRUE, message=FALSE,warning=FALSE}
### load require packages
library(dada2)
library(tidyverse)
library(Biostrings)

### setup working directory
#setwd("D:/System_bio/Microbiome/Manuscript/")   # CHANGE ME with the path where your fastq file is stored.
path <- "mock/"
Q30.path <- paste(path,"Q30/",sep="")
input.path <- list.files(Q30.path,pattern=".fastq",include.dirs = T,recursive = T,full.names = T)
nop.path <- paste0(path,"noprimers/", basename(input.path))
filter.path <- paste0(path,"noprimers/","filtered/")

```

```{r echo=TRUE, message=FALSE,warning=FALSE,eval=FALSE}
########  process sequencing files : Q30 filtered by seqkit #######

all.sample <- c("Zymo8.fastq","HM20e.fastq")
for(i in 1:length(all.sample)){
  sample.name <- gsub("(.*).fastq.*","\\1",all.sample[i],perl=T)
  Q30.filename <- paste(Q30.path,sample.name,".fastq",sep="")
  if(.Platform$OS.type == "windows"){
		shell(paste("seqkit seq -Q 30 ",path,all.sample[i]," > ",Q30.filename,sep=""))
	}else
		system(paste("seqkit seq -Q 30 ",path,all.sample[i]," > ",Q30.filename,sep=""))
}


```

## Run DADA2

The following workflow is recommended and demonstrated by *Benjamin J Callahan*. Please refer to <https://benjjneb.github.io/LRASManuscript/>.

```{r echo=TRUE, message=FALSE,warning=FALSE,eval=FALSE}
### 1. remove primer & orient sequences

F27 <- "AGRGTTYGATYMTGGCTCAG"
R1492 <- "RGYTACCTTGTTACGACTT"

prim <- removePrimers(input.path, nop.path, primer.fwd=F27, primer.rev=dada2:::rc(R1492), orient=TRUE, verbose=TRUE)

### 2. filter by expected errors
filTri <- filterAndTrim(nop.path, filter.path, minQ=3, minLen=1000, maxLen=1600, maxN=0, rm.phix=FALSE, maxEE=2)

### 3.Dereplicate
drp <- derepFastq(filter.path, verbose=TRUE)
### 4.Learn errors
err <- learnErrors(drp, BAND_SIZE=32, multithread=TRUE, errorEstimationFunction=dada2:::PacBioErrfun)
### 5.denoise
dd <- dada(drp, err=err, BAND_SIZE=32, multithread=TRUE)
### 6.Sequence table:construct an amplicon sequence variant table (ASV) table
seqT <- makeSequenceTable(dd)
#### 7. Check and Remove chimer
## If you're operating on a Linux system, you can enable the multithread option by setting it to TRUE.
bim <- isBimeraDenovo(seqT, minFoldParentOverAbundance=3.5, multithread=F)	
table(bim)

```

```{r echo=FALSE, message=FALSE,warning=FALSE}
load("20220224_mock.demo.RData")
table(bim)
```

##### One chimera needs to be removed.

```{r echo=TRUE, message=FALSE,warning=FALSE,eval=FALSE}
#### Remove chimer
## If you're operating on a Linux system, you can enable the multithread option by setting it to TRUE.
seqT.reBim <- removeBimeraDenovo(seqT, method = "per-sample", multithread=F, verbose=TRUE)	

#### 8. assign taxonomy 
DB <- "U16S/U16S.DADA2_format.fa"
## If you're operating on a Linux system, you can enable the multithread option by setting it to TRUE.
tax.reBim <- assignTaxonomy(seqT.reBim, DB, multithread=F,tryRC = F)
```

```{r echo=TRUE, message=FALSE,warning=FALSE}
## Get taxonomic assignment result of DADA2
taxsummary.reBim <- apply(unname(tax.reBim),1,function(x) paste(x, collapse=";")) 
taxsummary.reBim %>% head
```

```{r echo=TRUE, message=FALSE,warning=FALSE}
### extract abundance table
asv.abundance.reBim <- unname(t(seqT.reBim)) %>% data.frame
colnames(asv.abundance.reBim) <- gsub(".fastq","",row.names(seqT.reBim),perl=T)
asv.abundance.reBim %>% head

```

```{r echo=TRUE, message=FALSE,warning=FALSE}
## combine taxonomy and abundance
ASV.abu <- cbind(ASV = paste("ASV_",seq(nrow(asv.abundance.reBim)),sep=""), taxonomy=taxsummary.reBim, asv.abundance.reBim) 
ASV.abu %>% head


```

```{r echo=TRUE, message=FALSE,warning=FALSE}
### Get ASV sequences
ASV.seq <- DNAStringSet(getSequences(seqT.reBim))
names(ASV.seq) <- paste0("ASV_",seq(length(ASV.seq)))
print(ASV.seq)

```

```{r echo=TRUE, message=FALSE,warning=FALSE}
### pick up ASVs which were "NA" or indistinguishable species (including Shigella spp., Escherichia spp., Bacillus spp)
dada2.NA <- ASV.abu %>% filter(grepl("NA|Shigella|Escherichia|Bacillus",taxonomy))
dada2.NA %>% head

### export these sequence to perform BLAST
dada2.NA.seq <- ASV.seq[dada2.NA$ASV]
writeXStringSet(dada2.NA.seq,file="dada2.NA.ASV.fasta",format="fasta")


```

```{r echo=TRUE, message=FALSE,warning=FALSE}
### dada2.NA were perform by BLAST
blast.dir <- "e:/ncbi-blast-2.12.0+/bin/blastn.exe"
blast.subject <- "mock/U16S/U16S.blast_20230118.blastdb"
blast.query <- "dada2.NA.ASV.fasta"
blast.out <- "dada2.NA.ASV.blast.out"
blast_script <- paste(blast.dir," -task megablast -query ",blast.query," -db ",blast.subject," -out ",blast.out," -num_threads 20 -outfmt \"6 sseqid qseqid qlen length qstart qend sstart send pident evalue score qcovs\" -max_target_seqs 10",sep="")
shell(blast_script)

```

```{r echo=TRUE, message=FALSE,warning=FALSE}
################# process blast result
blast.result <- read.table(blast.out,sep="\t",header=F,fill=T,col.names=c("sseqid","qseqid","qlen","length","qstart","qend","sstart","send","pident","evalue","score","qcovs"))
blast.result %>% head

```

#### Dummy id was used in curated database. We can convert dummy id to taxonomy by blastIDmapping file.

```{r echo=TRUE, message=FALSE,warning=FALSE}
blast.IDmapping_file <- "mock/U16S/U16S.blastIDmapping.txt"
blast.IDmapping <- read.table(file=blast.IDmapping_file,header=T,sep="\t")
blast.result.2 <- merge(blast.result,blast.IDmapping[,c("rank.level","dummy.id")],by.x="sseqid",by.y="dummy.id",all.x=T)
tax.name.rank <- do.call(rbind,strsplit(blast.result.2$rank.level,";")) %>% as.data.frame
colnames(tax.name.rank) <- c("kingdom","phylum","class","order","family","genus","species")
blast.result.3 <- cbind(tax.name.rank,blast.result.2)
blast.result.3 %>% head

```

#### **The next step involves determining the reliability of the blast results.**

```{r echo=TRUE, message=FALSE,warning=FALSE}
source("DPOT_fun.R")
##### (a) Blast hits with qcovs >80 were retained; 
blast.result.q80 <- blast.result.3 %>% filter(qcovs > 80)
blast.result.list <- split(blast.result.q80,blast.result.q80$qseqid)

## If all hits pertain to an uncultured bacterium, they are preserved; however, any hit labeled as an 'uncultured bacterium' is removed if other specific bacteria are also identified.
blast.result.list.2 <- lapply(blast.result.list[unlist(lapply(blast.result.list,nrow))> 0],uncultured.check.fun)

##### (b) We retained, at most, the three top hits with  percent identity (pident) equal to or greater than the highest percent identity (HPID)reduced by 0.2, provided that the HPID was 99 or higher. 
##### (c) If HPID is less than 99, at most the top five hits with a percentage identity greater or equal to HPID minus 0.2 are retained.
blast.result.keep.pident <- lapply(blast.result.list.2,keep.pident.fun)

blast.result.keep.pident %>% head

```

```{r echo=TRUE, message=FALSE,warning=FALSE}
################# 3. integrate BLAST results into DADA2 assignTaxonomy
dada2.NA$taxonomy <- unlist(lapply(blast.result.keep.pident[dada2.NA$ASV],'[',1,"rank.level"))
mock.set <- rbind(ASV.abu %>% filter(!grepl("NA|Shigella|Escherichia|Bacillus",taxonomy)),dada2.NA)
mock.set %>% head

```

## Consolidated taxonomy

```{r echo=TRUE, message=FALSE,warning=FALSE, results='hold'}
mock.set.2 <- cbind(mock.set[,c("ASV","taxonomy")], sapply(mock.set %>% select(-c(ASV,taxonomy)),as.numeric))
mock.set.aggreate <- aggregate(.~taxonomy,mock.set.2 %>% select(-ASV),sum)
mock.set.aggreate %>% nrow	
mock.set.aggreate %>% head

```

```{r echo=TRUE, message=FALSE,warning=FALSE, results='hold'}
### split taxonomy into seven taxonomic ranks
tax.name.rank <- do.call(rbind,strsplit(mock.set.aggreate$taxonomy,";")) %>% as.data.frame
colnames(tax.name.rank) <- c("kingdom","phylum","class","order","family","genus","species")

mock.set.aggreate.2 <- cbind(mock.set.aggreate %>% select(-taxonomy),tax.name.rank) %>% as.data.frame
mock.set.aggreate.2 %>% head
```

### Verifying taxonomic assignment by DPOT with mock species

```{r echo=TRUE, message=FALSE,warning=FALSE}
## import the species contained in the mock set
mock.species <- read.table("mock/mock.species.txt",header=T,fill=T,quote = "",sep="\t")
mock.species %>% head

```

**column mock.sp : documented species contained in the mock set.** **column mock.sp.extend : mock species and those species have identical sequences**

### Verifying HM20e taxonomic assginment

```{r echo=TRUE, message=FALSE,warning=FALSE}
HM20e <- mock.set.aggreate.2 %>% filter(HM20e != 0) %>% select(-Zymo8)
HM20e$HM20e <- as.numeric(HM20e$HM20e)

### mock species list
answer.HM20e <- mock.species %>% filter(mock == "HM20e")
                  
HM20e$TP_FP <- ifelse(grepl(paste(answer.HM20e$mock.sp.extend, collapse="|"),HM20e$species),"TP","FP")
for(x in seq(nrow(answer.HM20e))){
  HM20e$mock.sp[grepl(answer.HM20e$mock.sp.extend[x],HM20e$species)] <- answer.HM20e$mock.sp[x]
}
HM20e %>% mutate(assgined.sp = species, document.sp = mock.sp) %>% select(TP_FP,HM20e,assgined.sp,document.sp)

```

According to the comparison table, all 20 species of the HM20e set were correctly assigned.

### Verifying Zymo8 taxonomic assginment

```{r echo=TRUE, message=FALSE,warning=FALSE}
Zymo8 <- mock.set.aggreate.2 %>% filter(Zymo8 != 0) %>% select(-HM20e)
Zymo8$Zymo8 <- as.numeric(Zymo8$Zymo8)

### mock species list
answer.Zymo8 <- mock.species %>% filter(mock == "Zymo8")
                  
Zymo8$TP_FP <- ifelse(grepl(paste(answer.Zymo8$mock.sp.extend, collapse="|"),Zymo8$species),"TP","FP")
for(x in seq(nrow(answer.Zymo8))){
  Zymo8$mock.sp[grepl(answer.Zymo8$mock.sp.extend[x],Zymo8$species)] <- answer.Zymo8$mock.sp[x]
}
Zymo8 %>% mutate(assgined.sp = species, document.sp = mock.sp) %>% select(TP_FP,Zymo8,assgined.sp,document.sp)

```

As per the comparison table, 10 species from the Zymo8 set were assigned. However, one species, *Bacillus intestinalis*, was misclassified. The Bacillus genus is known for its indistinguishable nature.
